### DISCUSSION TITLE
Hinting a known feasible solution to a MILP solver with Google OrTools mathopt in Python

### DISCUSSION THREAD

**Discussion by Sacropyne** [google/or-tools#4265] [Linear Solver questions]
Hi,

I have been going through the online documentation and the samples available on Github, but I was not able to find an example on how to provide an initial solution (or warm start, or solution hint, depending on the preferred synonym) to a compatible MILP solver when modelling through OrTools mathopt in Python.

Would you have a simple example available for this ?

Note if that helps that I am working with OrTools version 9.10.

Thanks a lot !

(I tried going through both the documentation and the source code, I found the SolutionHint class but I don’t know how to use it)

PS :  not being sure of the preferred channel for such a question I also posted on stackoverflow so here is the reference to the page :

https://stackoverflow.com/questions/78592175/hinting-a-known-feasible-solution-to-a-milp-solver-with-google-ortools-mathopt-i

**Comment by Mizux** [comment]
Looking for Solutionhint in ortools/math_opt/... src.

https://github.com/google/or-tools/blob/ef7bb07d57bbe73c5e2ff5bcff9894e0e3164f56/ortools/math_opt/core/python/solver_test.py#L191-L209
or
https://github.com/google/or-tools/blob/ef7bb07d57bbe73c5e2ff5bcff9894e0e3164f56/ortools/math_opt/python/model_parameters_test.py#L28-L38

**Reply by Sacropyne** [reply]
> Looking for Solutionhint in ortools/math_opt/... src.
> 
> https://github.com/google/or-tools/blob/ef7bb07d57bbe73c5e2ff5bcff9894e0e3164f56/ortools/math_opt/core/python/solver_test.py#L191-L209
> 
> 
> or
> https://github.com/google/or-tools/blob/ef7bb07d57bbe73c5e2ff5bcff9894e0e3164f56/ortools/math_opt/python/model_parameters_test.py#L28-L38

Thank you so much for your fast answer and sorry for not spotting these lines on my own.

I have tried to follow your second example, through this piece of code that is executed after a first optimization (to perform bi-criteria lexicographical optimization) :

```
# Initial solution
initial_solution_dictionary = dict()
for l in range(M):
    initial_solution_dictionary[all_vars[f"selection_{l+1}"]] = milp_result.variable_values(all_vars[f"selection_{l+1}"])
initial_solution = model_parameters.SolutionHint(initial_solution_dictionary)

# Second optim
milp.add_linear_constraint(all_vars[f"min"] >= milp_result.objective_value())

# Criterion : minimization of the number of selected slats
milp.maximize((1/(N*(N-1)/2))*(mathopt.fast_sum([all_vars[f"distance_{l+1}_{ll+1}"] for l in range(M) for ll in range(M) if ll>l])-distances.max()*(M*(M-1)/2-N*(N-1)/2)))

# Launch
params = mathopt.SolveParameters(enable_output=True)
model_params = model_parameters.ModelSolveParameters()
model_params.solution_hints.append(initial_solution)
milp_result = mathopt.solve(milp, mathopt.SolverType.HIGHS, params=params,model_params=model_params)
```

It runs without error but it looks like Highs is insensitive to the provided hint, it behaves just like when it’s not there. Is it maybe that the functionality is not there yet for Highs through mathopt ? Or am I still missing something in how I should use these classes and methods ?

**Comment by rma350** [comment]
If you look in this file:

https://github.com/google/or-tools/blob/stable/ortools/math_opt/solvers/highs_solver.cc

You see that hint is not supported for HiGHS yet (it is just silently ignored).

Gurobi, CP-SAT, and SCIP all support hint if this feature is critical. Gurobi and SCIP also support incremental MIP solving, which might buy you a little more than the hint if you are just changing the objective and resolving. Gurobi also has its own hierarchial objectives which we have limitted support for.
